cumulus/primitives/parachain-inherent/src/lib.rs

```diff
@@ -85,6 +88,276 @@ pub mod v0 {
+	messages: Vec<Message>,
+
+	_phantom: core::marker::PhantomData<BlockNumber>,
+}
+
+impl<BlockNumber: PartialEq + PartialOrd, Message: InboundMessage<BlockNumber>>
+	InboundMessagesCollection<BlockNumber, Message>
+{
+	pub fn new(messages: Vec<Message>) -> Self {
+		Self { messages, _phantom: Default::default() }
+	}
+
+	/// Drop all the messages up to `last_processed_msg`.
+	pub fn drop_processed_messages(&mut self, last_processed_msg: &InboundMessageId<BlockNumber>) {
+		let mut last_processed_msg_idx = None;
+		let messages = &mut self.messages;
+		for (rev_idx, message) in messages.iter().rev().enumerate() {
+			let idx = (messages.len() - rev_idx - 1) as u32;
+			let sent_at = message.sent_at();
+			if sent_at == last_processed_msg.sent_at {
+				last_processed_msg_idx = idx.checked_sub(last_processed_msg.reverse_idx);
+				break;
+			}
+			if sent_at < last_processed_msg.sent_at {
+				last_processed_msg_idx = Some(idx);
+				break;
+			}
+		}
+		if let Some(last_processed_msg_idx) = last_processed_msg_idx {
+			messages.drain(..last_processed_msg_idx as usize + 1);
+		}
+	}
+
+	pub fn into_compressed(
+		self,
+		size_limit: &mut usize,
+	) -> CompressedInboundMessagesCollection<BlockNumber, Message> {
+		let mut messages = self.messages;
+
+		let mut maybe_split_off_pos = None;
+		for (idx, message) in messages.iter().enumerate() {
+			if *size_limit < message.data().len() {
+				break;
+			}
+			*size_limit -= message.data().len();
+
+			maybe_split_off_pos = idx.checked_add(1);
+		}
+
+		let mut compressed_messages = vec![];
+		if let Some(split_off_pos) = maybe_split_off_pos {
+			let extra_messages = messages.split_off(split_off_pos);
+			compressed_messages = extra_messages.iter().map(|msg| msg.to_compressed()).collect();
+		}
+
+		CompressedInboundMessagesCollection {
+			uncompressed_messages: messages,
+			hashed_messages: compressed_messages,
+			_phantom: Default::default(),
+		}
+	}
+}
+
+/// A compressed collection of inbound messages.
+///
+/// The first messages in the collection (up to a limit) contain the full message data.
+/// The messages that exceed that limit are hashed.
+#[derive(
+	codec::Encode,
+	codec::Decode,
+	codec::DecodeWithMemTracking,
+	sp_core::RuntimeDebug,
+	Clone,
+	PartialEq,
+	TypeInfo,
+)]
+pub struct CompressedInboundMessagesCollection<BlockNumber, Message: InboundMessage<BlockNumber>> {
+	uncompressed_messages: Vec<Message>,
+	hashed_messages: Vec<Message::CompressedMessage>,
+
+	_phantom: core::marker::PhantomData<BlockNumber>,
+}
+
+impl<BlockNumber, Message: InboundMessage<BlockNumber>>
+	CompressedInboundMessagesCollection<BlockNumber, Message>
+{
+	pub fn messages(&self) -> (&[Message], &[Message::CompressedMessage]) {
+		(&self.uncompressed_messages, &self.hashed_messages)
+	}
+
+	pub fn check_advancement_rule(&self, collection_name: &str) {
+		if self.uncompressed_messages.len() == 0 && self.hashed_messages.len() > 0 {
```

Comment:
We need to check that uncompressed_messages is as the maximum allowed size when there is any hashed message. Or in other words, there can not be any hashed message if we did not reached the maximum allowed size.

Reply:
Maybe not the full maximum allowed size, but we have some maximum message size. So, there should be more than 1 etc.